### PR TITLE
Respect minimumHealthCapacity during leader election

### PR DESCRIPTION
**Pull request by kamaradclimber**
Before this patch, during a deployment, a leader election would lead
marathon to not respect minimumHealthCapacity parameter.
Reason is that TaskReplaceActor used to ignore instance healthyness when
considering instance that could be killed "immediately" upon actor
start.

We now respect minimumHealthCapacity by taking into account healthy
instances.

This patch required to add an property "this app has configured
healthchecks" to be able to distinguish following cases:
- app has HC but we don't have signal yet for given instance
- app has HC and we have signal of healthyness for given instance
- app has no HC

It also brings ability to fix TODO at
https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/core/instance/Instance.scala#L267
in a future patch since we now know difference between no HC and no
information about HC.

Fixes: MARATHON-8716
Change-Id: Ia7b11cbb22f86967b49298f774c6d27fc01a6e58
Signed-off-by: Grégoire Seux <g.seux@criteo.com>

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala [hunk 1]
```diff
@@ -51,8 +51,11 @@ class TaskReplaceActor(
   // Old and new instances that have the Goal.Running
   val activeInstances = currentInstances.filter(_.state.goal == Goal.Running)
 
+  // Active instances that are healthy (or don't have health checks)
+  val healthyOrNoHealthCheckInstances = currentInstances.filter((i) => !i.hasConfiguredHealthChecks || i.state.healthy.getOrElse(false))
+
   // The ignition strategy for this run specification
-  private[this] val ignitionStrategy = computeRestartStrategy(runSpec, activeInstances.size)
+  private[this] val ignitionStrategy = computeRestartStrategy(runSpec, healthyOrNoHealthCheckInstances.size)
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: for the method computeRestartStrategy lets change the name of the parameter from `runningInstancesCount` to `consideredHealthyCount`

**Comment by kamaradclimber** [comment]
Thanks @timcharper for your suggestions, I've implemented them and repushed

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala [hunk 2]
```diff
@@ -285,20 +285,20 @@ object TaskReplaceActor extends StrictLogging {
   /** Encapsulates the logic how to get a Restart going */
   private[impl] case class RestartStrategy(nrToKillImmediately: Int, maxCapacity: Int)
 
-  private[impl] def computeRestartStrategy(runSpec: RunSpec, runningInstancesCount: Int): RestartStrategy = {
+  private[impl] def computeRestartStrategy(runSpec: RunSpec, consideredHealthyInstancesCount: Int): RestartStrategy = {
     // in addition to a spec which passed validation, we require:
     require(runSpec.instances > 0, s"instances must be > 0 but is ${runSpec.instances}")
-    require(runningInstancesCount >= 0, s"running instances count must be >=0 but is $runningInstancesCount")
+    require(consideredHealthyInstancesCount >= 0, s"running instances count must be >=0 but is $consideredHealthyInstancesCount")
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Hmm, with this change it's possible that all instances are unhealthy and this actor will crash now. We'll need to handle zero appropriately. @jeschkies can you weigh in?

Comment (kamaradclimber) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: You're right.
From what I understand, this requirement was a safety because 0 running instances meant there was no task to replace (only task to start, which is handled by a different deployment step).

We can now have 0 healthy instance but >0 running instances.

I would naively tend to remove that assertion (following code will lead to creation of a RestartStrategy with nrToKillImmediately = 0).
Tasks from old app version will be killed later as we received healthy instance messages.

What do you think?

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Hmpf, this good is the most brittle we have in all of Marathon. @kamaradclimber, could you give an example with let's say three instances to illustrate the issue you've encountered. Cause I don't quite understood the issue your trying to fix.

Comment (kamaradclimber) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Sure!
 
Let's assume an application with 3 instances. Instances take 15minutes to pass their healthchecks.
Application configuration has a minimumHealthCapacity of 0.6 (allowing 1 instance to be killed by marathon during deployments) and maximumOverCapacity of 0 (never go beyond 3 instances).

- T+0m: a deployment of the application is initiated. We have 3 instances: A1, B1, C1
- T+0m: marathon kills one instance (A1), start a new one (A2) which is not healthy
- T+5m: marathon leader disappear, a new marathon leader is elected
- T+5m: new marathon leader kills B1, spawns B2.

At this moment, minimumHealthCapacity is violated because state is the following: A2 (not healthy), B2 (not healthy), C1 (healthy).

Later (at T+15m), A2 will pass its healthchecks and deployment will continue.

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Thanks. That helps a lot. So, it seems we do not reconcile for old instance but only already started instances. See `TaskReplaceActor:201` and `ReadinessBehaviour:195`. I actually came across this recently and was not sure which entity actually sets `instance.state.healthy`. We have Mesos and Marathon health checks. I know for sure that Mesos health checks alter the `instance.state.healthy` state. However, I don't know if Marathon health checks do the same. If they do we could simplify `ReadinessBehaviour` and the ignition strategy. Maybe that's too much for now.

Just so that we are not altering the logic to much. What do you think about moving the health filtering to the ignition strategy? Pass the seq instead of an integer.

Comment (kamaradclimber) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Thanks for your input.
I haven't thought about mesos health checks vs marathon health checks. Two pointers make me think this is ok:
- when testing this patch for real, I've used marathon health checks (so they are taken in consideration by instance.state.healthy)
- I've followed code path to confirm mesos health checks are indeed used to update `instance.state.healthy`

I'm not so sure about changing signature of `computeRestartStrategy` method, a lot of tests in RestartStrategyTest.scala file would need to be modified (they currently pass an int to the method and would need to have a list of healthy instances to be stubbed).
Any specific reason to move health filtering logic in this method?

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: > Any specific reason to move health filtering logic in this method?
Yes. A lot of the other logic depends on `activeInstances`. I would rather not pull in the health checks there. If we would receive updates for health changes and have it attached to the instances it would be a different story.

Comment (kamaradclimber) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: I'm not sure to see the logic depending on `activeInstances` (apart from test).

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Oh, you are right. I mixed it up with `instancesAlreadyStarted`.

**Comment by jeschkies** [comment]
There are quite a few unused imports
```
[warn] /home/admin/workspace/marathon-pipelines_PR-7115/src/test/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActorTest.scala:20:43: Unused import
[warn] import mesosphere.marathon.core.task.Task.Status
[warn]                                           ^
[warn] /home/admin/workspace/marathon-pipelines_PR-7115/src/test/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActorTest.scala:21:44: Unused import
[warn] import mesosphere.marathon.core.task.state.NetworkInfo
[warn]                                            ^
[warn] /home/admin/workspace/marathon-pipelines_PR-7115/src/test/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActorTest.scala:28:33: Unused import
[warn] import org.apache.mesos.Protos.{CheckStatusInfo, TaskID, TaskState, TaskStatus}
[warn]                                 ^
[warn] /home/admin/workspace/marathon-pipelines_PR-7115/src/test/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActorTest.scala:28:50: Unused import
[warn] import org.apache.mesos.Protos.{CheckStatusInfo, TaskID, TaskState, TaskStatus}
[warn]                                                  ^
[warn] /home/admin/workspace/marathon-pipelines_PR-7115/src/test/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActorTest.scala:28:58: Unused import
[warn] import org.apache.mesos.Protos.{CheckStatusInfo, TaskID, TaskState, TaskStatus}
[warn]                                                          ^
[warn] /home/admin/workspace/marathon-pipelines_PR-7115/src/test/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActorTest.scala:28:69: Unused import
[warn] import org.apache.mesos.Protos.{CheckStatusInfo, TaskID, TaskState, TaskStatus}
[warn]                                                                     ^
[warn] /home/admin/workspace/marathon-pipelines_PR-7115/src/test/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActorTest.scala:29:55: Unused import
[warn] import org.scalatest.concurrent.PatienceConfiguration.Timeout
[warn]     
```

That's why the build fails.

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala (hunk 2 above)
Comment (kamaradclimber) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Could we list remaining actions to do on this PR, I'm a bit lost.

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: > Could we list remaining actions to do on this PR, I'm a bit lost.

It's all good from my side. All you need to do is to remove the unused imports I've listed below so that the build passes.

**Comment by kamaradclimber** [comment]
> There are quite a few unused imports

good catch, thanks. I've pushed a fix. I'm using `sbt universal:packageZipTarball` to build, what was the command to see those warnings? (not familiar at all with sbt)

**Comment by jeschkies** [comment]
> good catch, thanks. I've pushed a fix. I'm using sbt universal:packageZipTarball to build, what was the command to see those warnings? (not familiar at all with sbt)

Unfortunately we could not enable an error for unused imports unless all warnings should be errors. That's why we simply check the logs after the build https://github.com/mesosphere/marathon/blob/master/ci/pipeline#L77. This is super dirty.

**Comment by kamaradclimber** [comment]
Let me know if I can fix anything else

**Comment by kamaradclimber** [comment]
Anyone to give me the error (or how to reproduce it)?

**Comment by jeschkies** [comment]
The following test fails
`mesosphere.marathon.integration.GroupDeployIntegrationTest.GroupDeployment should An upgrade in progress cannot be interrupted without force`

Here is the log file [PR-7115-6.log.tar.gz](https://github.com/mesosphere/marathon/files/4108418/PR-7115-6.log.tar.gz)

**Comment by kamaradclimber** [comment]
Thanks, I'll work to reproduce this failing test on my machine and fix it.

**Comment by kamaradclimber** [comment]
I think I understand why this test is failing.
Setup is pretty simple: app has 1 instance, 1 health check

Scenario is the following:
- v1 of the app is deployed and passes its health check
- a buggy v2 is deployed (not passing its health check) => deployment never finishes
- a v3 is deployed without force => triggers a conflict
- a v4 is deployed with force (and without bug)

Marathon is facing the following conditions for deployment of v4:
- 1 healthy task in v1
- 1 non healthy task in v2

minimumHealthCapacity is 1 (default value) and maximumOverCapacity is 1. Number of instances defined in configuration is 1. It allows marathon to move between 1 and 2 instances.

Our new code says we cannot kill any instance immediately (because we have 1 healthy instance and we cannot go below 1). We cannot spawn a new one either because of the maximumOverCapacity.

Even if we allowed marathon to kill one instance (what use to happen before my patch), the killed one might be the only healthy instance.
I think it should systematically prefer killing unhealthy instances (of the old version).

I'll try to find a solution to those issues.

**Comment by kamaradclimber** [comment]
I've implemented a first version to correct this behavior. Feedback is welcomed (I'm running tests on my machine to see if anything breaks).

**Comment by kamaradclimber** [comment]
Tests are now passing on my side.
To be precise I have exactly the same tests failing on my machine before and after my patches:
```
[error] Error: Total 148, Failed 3, Errors 5, Passed 140
[error] Failed tests:
[error] 	mesosphere.marathon.integration.AppDeployIntegrationTest
[error] 	mesosphere.marathon.integration.ResidentTaskIntegrationTest
[error] 	mesosphere.marathon.integration.GpuSchedulingIntegrationTest
[error] Error during tests:
[error] 	mesosphere.marathon.integration.SharedMemoryIntegrationTest
[error] 	mesosphere.marathon.integration.SeccompIntegrationTest
[error] 	mesosphere.marathon.integration.DockerAppIntegrationTest
[error] 	mesosphere.marathon.integration.UpgradeIntegrationTest
[error] 	mesosphere.marathon.integration.MesosAppIntegrationTest
```

**Comment by jeschkies** [comment]
@timcharper could you take another look? @kamaradclimber, the tests that failed on your machine only run on Linux.

**Comment by kamaradclimber** [comment]
Thanks you both for your reviews. What should I do to get it merged?

**Comment by jeschkies** [comment]
@kamaradclimber, nothing. I landed it. Thanks for you endurance.

**Comment by kamaradclimber** [comment]
Thanks you both for your advices. My team mates are likely to submit
another patch soon  for a similar but we have observed in production.

**Comment by jeschkies** [comment]
> My team mates are likely to submit another patch soon  for a similar but we have observed in production.

This is great news! Which version of Marathon are you running?

**Comment by kamaradclimber** [comment]
We are still running 1.6.x (with a few custom patches to support network bandwidth resource and a safety feature to avoid killing too many instances if all instances of an app stop passing their healthchecks)

**Comment by kamaradclimber** [comment]
It seems this PR only works for applications with mesos health checks only.

A simple application with only marathon health checks and maximumOverCapacity set to 0 cannot be updated anymore with this patch.

To reproduce:
- app has a simple healthcheck, 10 instances
- maximumOverCapacity 0
- minimumHealthCapacity is set to any value > 0 (0.5 for instance)

When taskReplaceActor starts, it looks at instanceTracker to count healthy instances. It seems that instanceTracker is not aware of instance health checks status (for marathon health checks) so no instance is considered healthy.
So taskReplaceActor does not kill any task "immediately" and is blocked forever.

**Comment by kamaradclimber** [comment]
I suggest to revert my PR since the fix for this bug does not seem obvious to me. I'm thinking about this but any guidance would be appreciated.

(I can also open a proper ticket if you prefer)

**Comment by kamaradclimber** [comment]
After discussion with my team mates here are how I understand the issue.

My PR initially used instanceTracker to get list of instance (and eventually their health status). Instance tracker only knows about mesos status.
Only way to know health status of instances (with marathon healthchecks) is to be an actor and subscribe to health events.

There are two scenarios that we should handle.
We consider in both case an app with maximumOverCapacity at 0 and 1 marathon health check.

Scenario 1:
- marathon leader has been running for a while
- app has been running for a while, all instances have a known health status
- a new version of app is being deployed

Scenario 2:
- a new version of app is being deployed
- marathon leader just crashed and a new one is taking over

To cope with scenario 1, my suggestion is for the TaskReplaceActor to send a message to another actor (the HealthCheckActor) to ask for a list of healthy instances. Based on that information and the instancetracker (to deal with mesos healthchecks), we can decide how many instances to kill immediately.

Scenario 2 introduces a additional challenge since, at leader startup, healthcheckactor has not a complete view of all instance health status.
Proposition for scenario1 would fall short since we would have a decision not to kill any instance.
To cope with this, I propose the TaskReplaceActor to _frequently_ ask HealthCheckActor healthy instances count and launch instance killing based on that information.

This would also allow to deal with another bug (not yet reported) that was present in marathon (before my PR) due to https://github.com/mesosphere/marathon/blob/v1.8.232/src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala#L194. It this line, we kill an old instance whenever we receive event about another instance being ready/healthy.
This can be dangerous (put the app below minimumHealthCapacity) for two reasons:
- another task might have died for any other reason (and its eventual replacement not being started/healthy yet)
- we can receive the event of a given task being healthy several time (that's the condition we've observed on production)

To summarize, the TaskReplaceActor should have the following behavior

- at startup: do nothing
- frequently (at frequent interval and upon task healthy event): send a message to HealthCheckActor to get list of healthy instance, based on that information kill some tasks if it is safe.

What do you think?

**Comment by jeschkies** [comment]
> To cope with scenario 1, my suggestion is for the TaskReplaceActor to send a message to another actor (the HealthCheckActor) to ask for a list of healthy instances.

I really do not like that we keep the health results in a separate actor. Ideally the instances would be update in the instance tracker with their health results. This would simplify things quite a bit.
